### PR TITLE
Roof

### PR DESCRIPTION
**Pull request by jonathan2222**
Added a roof to each room.

### CODE REVIEW

File: GameProject/Engine/Rendering/Pipeline.cpp [hunk 1]
```diff
@@ -415,16 +417,18 @@ void Pipeline::calcDirLightDepthInstanced(const std::vector<std::pair<RenderingT
...
 			drawModelPrePassInstanced(pair.first.model);
 	}
+	glCullFace(GL_BACK);
 
 	this->ZprePassShaderInstanced->unbind();
 	this->prePassDepthOff();
 	this->shadowFbo.unbind();
 
-/*#ifdef IMGUI
+#ifdef IMGUI
```
Comment (TheoBerlin) [inline GameProject/Engine/Rendering/Pipeline.cpp]: Comment this out again.

File: GameProject/Engine/Rendering/Pipeline.cpp [hunk 2]
```diff
@@ -482,7 +486,8 @@ void Pipeline::addCurrentLightManager(LightManager * lm)
 	this->entityShaders[DRONE_SHADER]->updateLightMatrixData(lightManager->getLightMatrixPointer());
 	this->entityShaders[WALL]->updateLightMatrixData(lightManager->getLightMatrixPointer());
 	this->entityShaders[INFINITY_PLANE]->updateLightMatrixData(lightManager->getLightMatrixPointer());
-	this->entityShaders[INFINITY_PLANE_PREPASS]->updateLightMatrixData(lightManager->getLightMatrixPointer());
+	this->entityShaders[INFINITY_PLANE_PREPASS]->updateLightMatrixData(lightManager->getLightMatrixPointer());	
+	//this->entityShaders[ROOF_PLANE]->updateLightMatrixData(lightManager->getLightMatrixPointer());
```
[the remark this answers is not in the document]
Comment (jonathan2222) [reply GameProject/Engine/Rendering/Pipeline.cpp]: Done.

File: GameProject/Engine/Rendering/Shaders/InfinityPlanePrePassShader.frag [hunk 3]
```diff
@@ -21,8 +21,8 @@ bool InRoom(vec3 p, int size, int index)
         // Check if the horizontal line, which p.z spans, is intersecting the line between w1 and w2.
         if((w2.z < p.z && w1.z >= p.z) || (w1.z < p.z && w2.z >= p.z)) {
             // Check if the point is not the left or the right side of the line.
-            if(w2.x + (w1.x-w2.x)*(p.z + w2.z)/(w1.z-w2.z) < p.x) {
-                isOdd = !isOdd;
+            if(((w1.x-w2.x)/(w1.z-w2.z)) * (p.z - w1.z) + w1.x > p.x) {
+				isOdd = !isOdd;
```
[the remark this answers is not in the document]
Comment (jonathan2222) [reply GameProject/Engine/Rendering/Shaders/InfinityPlanePrePassShader.frag]: Done.

File: GameProject/Engine/Rendering/Shaders/RoofShader.frag [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+
+void main()
+{
+    bool isBorderX = bool(step(cellSize-borderWidth, mod(fragPos.x, cellSize)));
+    bool isBorderZ = bool(step(cellSize-borderWidth, mod(fragPos.z, cellSize)));
+    float l =  1.0 - float(isBorderZ || isBorderX);
+
+    float minDist = 10000.0;
+	int index = 0;
+    for (int i = 0; i < wallPoints.groupSize; i++)
```
[the remark this answers is not in the document]
Comment (jonathan2222) [reply GameProject/Engine/Rendering/Shaders/RoofShader.frag]: Done.

File: GameProject/Engine/Rendering/Pipeline.cpp (hunk 1 above)
Comment (jonathan2222) [reply GameProject/Engine/Rendering/Pipeline.cpp]: Done.

**Pull request merged**